### PR TITLE
Add property for opting out of apphost signing in macOS

### PR DESCRIPTION
**Pull request by mateoatr**
Add support for the new `CreateAppHost` argument, `enableMacOSCodeSign`. This PR adds property `_EnableMacOSCodeSign` to allow users to opt-out of apphost signing in macOS.

Dependent upon https://github.com/dotnet/runtime/pull/53913.

### CODE REVIEW

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets
Comment (vitek-karas) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets:512]: What happens if I try to build a linux binary on mac? (Obviously we must not try to sign it)

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets
Comment (vitek-karas) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets:123]: While I support this change I think we should make the change more visible. I don't know what's the process for this - technically it could be seen as a mild breaking change. @agocke or @marcpopMSFT  might know how we track these. Short version:
* .NET 6 SDK will generate apphost executable for macOS apps by default for all TFMs that support apphost
* It will now codesign it with the default cert when running on macOS
* In .NET 5 we disabled apphosts on macOS because we didn't sign them and default settings OS would not run such executables.

(This will help a lot with making things like `dotnet run` work on Apple M1 when targetting downlevel TFMs)

Comment (marcpopMSFT) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets:123]: Minor changes can go into release notes but full breaking changes get filed as issues in the doc's repo to go into the breaking change documentation.  Since it's a change in default behavior, I'd probably lean towards breaking changes so filing an issue with the details using this link: https://github.com/dotnet/docs/issues/new?assignees=&labels=&template=dotnet-breaking-change.md&title=

Comment (agocke) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets:123]: I agree, this is probably not a breaking change for most people, but it seems good to have it documented as such, just in case.

Comment (mateoatr) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets:123]: Opened the breaking change issue: https://github.com/dotnet/docs/issues/24766

Comment (vitek-karas) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets:123]: There is another issue with this change unfortunately:
On Apple M1 when running with native SDK (so arm64), if I try to build a 3.1 app, it would:
* Try to generate apphost (as per the above change)
* It would imply implicit RID which is the SDK's RID, which would be `osx-arm64`

This will FAIL - there's no osx-arm64. It's possible to try this today:
```
> dotnet new console -f netcoreapp3.1
> dotnet publish -r osx-arm64 --self-contained false
error NETSDK1084: There is no application host available for the specified RuntimeIdentifier 'osx-arm64'.
```

Related to the discussion here: https://gist.github.com/richlander/8cd7b4570775d40f9024561883d79516#overriding-the-implict-rid

@richlander for context

**Review by vitek-karas — CHANGES_REQUESTED** [review]
I think we will have to somehow address the implicit RID on Apple M1 problem before merging this, as it would make what is a very mild breaking change a very hard break (`dotnet build` doesn't work by default on 3.1 app on osx-arm64 SDK).

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets
Comment (vitek-karas) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets:68]: @wli3 or @dsplaisted could you please review this piece? The goal is when running on osx-arm64 and working with downlevel app (TFM < 6.0) to switch to osx-x64 since the first version which supports osx-arm64 is .NET 6 (downlevel versions don't have support and only run on x64, also macOS arm64 can emulate x64, so it's the right thing to do to run the app as x64).

The main concern we had was how to construct the new RID - specifically if it should maintain the potential OS version and such in the RID or not.

**Review by vitek-karas — APPROVED** [review]
Looks good to me, but this should also be reviewed by somebody from SDK to validate mainly the implicit RID calculation.

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets
Comment (dsplaisted) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets:68]: This will only apply for projects that opt in to `UseCurrentRuntimeIdentifier`, or when `--use-current-runtime` is passed on the command line.  Is that the intention?

I thought that we were going to use this type of logic for the the `DefaultAppHostRuntimeIdentifier`.

It's not great that we have to do string manipulation to check the RuntimeIdentifier, but I don't think there's a better option.

I do think when setting the value you don't need to use `String.Copy`.  You can just call replace directly, String.Copy is a workaround to call replace on item metadata, and shouldn't be needed for a property value.

Comment (vitek-karas) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets:68]: I don't know what's the difference between `UseCurrentRuntimeIdentifier` and `DefaultAppHostRuntimeIdentifier`. The intent of this change is that typing `dotnet run` on a 3.1 project when running on macOS arm64 SDK will produce macOS x64 app output (apphost) and thus will be able to execute the app. This then also means that `dotnet build` should behave that way.

In short - SDK should not default to the RID of the SDK when targeting downlevel TFMs and such TFMs are not supported on the RID. It should default to whatever RID is actually supported for the TFM on the current machine.

Comment (dsplaisted) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets:68]: `DefaultAppHostRuntimeIdentifier` is the primary property you should be concerned with.  That's what determines which apphost gets used if a `RuntimeIdentifier` (or `AppHostRuntimeIdentifier`) isn't otherwise specified.

`UseCurrentRuntimeIdentifier` is a new-ish option to allow building a self-contained app without explicitly specifying the RuntimeIdentifier.  See #14093.  It is probably a good idea for `UseCurrentRuntimeIdentifier` to switch from arm64 to x64 when targeting .NET 5.0 or earlier, but that's not the mainline scenario you're trying to address.

Comment (mateoatr) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets]: That makes sense. I've changed things so we modify the `DefaultAppHostRuntimeIdentifier`.

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets
Comment (agocke) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets:136]: Do we need `EndsWith('arm64')`? Isn't it good enough to say `osx`+`<6.0`?

And why use `NETCoreSdkRuntimeIdentifier` instead of `DefaultAppHostRuntimeIdentifier`?

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets
Comment (agocke) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets]: If `UseCurrentRuntimeIdentifier` isn't set to `DefaultAppHostRuntimeIdentifier` I think we should set that too. Otherwise I think that scenario will be broken for Mac.

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets
Comment (mateoatr) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets]: Changed to use `DefaultAppHostRuntimeIdentifier` instead of `NETCoreSdkRuntimeIdentifier`. I don't think we need the `Ends...` here, it's enough to check for `osx & < 6.0`, but left it here to make things clearer (since we are replacing that same string later).

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets
Comment (marcpopMSFT) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets:121]: To clarify per Teams chat, the plan is for the implicit RID when targeting downlevel runtimes when using the arm64 SDK is to cross-target to x64 for emulation.

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets
Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets:138]: I think it would be good to have a comment explaining the reasoning behind this logic.

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets
Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets:512]: For framework-dependent apps, `RuntimeIdentifier` won't be set.  Don't you still want to sign the apphost in that case?

File: src/Tests/Microsoft.NET.Build.Tests/AppHostTests.cs
Comment (dsplaisted) [inline src/Tests/Microsoft.NET.Build.Tests/AppHostTests.cs:39]: I think we should use NuGet TargetFramework parsing APIs (`NuGetFramework.Parse` I believe) here instead of ad-hoc manual parsing.

File: src/Tests/Microsoft.NET.Build.Tests/AppHostTests.cs
Comment (dsplaisted) [inline src/Tests/Microsoft.NET.Build.Tests/AppHostTests.cs:119]: I'd suggest using `RunExeCommand` here, which handles running the process and capturing the output.

File: src/Tests/Microsoft.NET.Build.Tests/AppHostTests.cs
Comment (dsplaisted) [inline src/Tests/Microsoft.NET.Build.Tests/AppHostTests.cs:154]: Also here for using `RunExeCommand`

File: src/Tests/Microsoft.NET.TestFramework/TestAsset.cs
Comment (dsplaisted) [inline src/Tests/Microsoft.NET.TestFramework/TestAsset.cs:100]: Most of our test assets don't set the `TargetFramework` via `WithTargetFramework`.  So this new property would only be set for a small portion of our tests, while it looks like it should work generally.  Can you try to find another way to solve the problem?  For example, you could add a `targetFramewokr` parameter to the `GetExpectedFilesFromBuild` method.

**Review by dsplaisted — COMMENTED** [review]
There are tests that verify that we don't sign the apphost, but it doesn't look like there are any that verify that the apphost is signed correctly.  It also looks like it's not currently being signed in the framework-dependent case.

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets
Comment (vitek-karas) [reply src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets]: We definitely want to sign the apphost in all cases

**Comment by lewing** [comment]
This is approved and green, is there something that needs additional review?

**Comment by vitek-karas** [comment]
We need sign-off from somebody from SDK (for example Daniel).

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets
Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.RuntimeIdentifierInference.targets:139]: Use `[MSBuild]::VersionLessThan` here instead of `&lt;` for version comparison.

File: src/Tests/Microsoft.NET.Build.Tests/AppHostTests.cs
Comment (dsplaisted) [inline src/Tests/Microsoft.NET.Build.Tests/AppHostTests.cs:41]: The `Version` type has somewhat unexpected comparison rules (5.0 is not the same as 5.0.0.0), so I would suggest just comparing against the `Major` version.
```suggestion
                if (parsedTargetFramework.Version.Major >= 5)
```

File: src/Tests/Microsoft.NET.Build.Tests/AppHostTests.cs
Comment (dsplaisted) [inline src/Tests/Microsoft.NET.Build.Tests/AppHostTests.cs:44]:
```suggestion
                if (parsedTargetFramework.Version.Major < 6)
```

File: src/Tests/Microsoft.NET.Build.Tests/AppHostTests.cs
Comment (dsplaisted) [inline src/Tests/Microsoft.NET.Build.Tests/AppHostTests.cs:55]: Now that this runs on macOS, can this just be a `Theory` instead of a `PlatformSpecificTheory`?

**Review by dsplaisted — APPROVED** [review]
A few comments, but looks good.

**Pull request merged**
